docs: mcp (#85010)
vercel/next.js: docs/01-app/02-guides/mcp.mdx

---
title: Enabling Next.js MCP Server for Coding Agents
nav_title: Next.js MCP Server

> description: Learn how to enable the experimental Next.js MCP server to allow agents access your application state

Reviewer: > description: Learn how to enable the experimental Next.js MCP server to allow agents access to your application state

---

> The Model Context Protocol (MCP) is an open standard that allows AI agents and coding assistants to interact with your applications through a standardized interface. The experimental Next.js MCP server enables agents to understand and work with your application's structure and state in real-time.

Reviewer: needs a link to the MCP protocol docs

> When combined with [`next-devtools-mcp`](https://www.npmjs.com/package/next-devtools-mcp), your agent gets access to comprehensive Next.js development tools including documentation, browser testing capabilities, Chrome DevTools functionality, and automated MCP tools for common development tasks.

Reviewer: what is the chrome devtools functionality? needs a link to their MCP

> ## Enabling the MCP server

Reviewer: what's unclear to me is what does the MCP server provides on its own? if nothing, let's not even mention it and let's say that you HAVE to use the devtools MCP

Author: We need both, the mcp inside next.js is exposing internals, the `next-devtools-mcp` is for managing your next.js apps and providing other functionalities like query docs, cache components guide and migration, upgrade next 16. We recommend to have it both. Eventually we want to remove the mcpServer flag in next, let it either enable by default or opt-in when you use the package. so far you need to enable it yourself

Reviewer: can we can not have a nicer DX and automatically launch/relaunch the next server when we detect that the next.js devtools MCP is active?

Author: There's a trickt builtin next-devtools-mcp that you can enable dev server with a env var to opt-in enabling next mcp, e.g. when doing cache components migration. But in daily basis development you'd better to enable it manually most of the time. 

We want to have this always enabled later once this flow becomes stable, so you don't need to have the flag.

Author: Removed mentioning mcpServer flag as it's enabled by default now.

With `experimental.mcpServer` enabled and `next-devtools-mcp` configured, agents get access to:

- **Next.js Knowledge Base** - Comprehensive Next.js documentation and best practices
- **Debugging Insights** - Real-time error detection, build diagnostics, logs, and runtime information from your dev server
- **Migration Tools** - Automated helpers for migrating your application, including upgrading to Next.js 16 with codemods and enabling Cache Components with automated setup and error detection

## Using with agents

When you have the MCP server enabled, MCP-compatible Coding Agents can automatically discover and connect to your Next.js application with the experimental `mcpServer` flag enabled. This provides agents with rich context about your project.

### Benefits for agent-assisted development

With the MCP server enabled, agents can:

- **Understand your project layout** - Know exactly where routes, components, and utilities are located
- **Make context-aware suggestions** - Recommend the right place to add new features based on your existing structure
- **Query live application state** - Check current configuration, routes, and middleware during development

> - **Reduce context overhead** - No need to manually explain your app structure each conversation

Reviewer: that's the same point as 1

> - **Provide accurate implementations** - Generate code that follows your project's patterns and conventions

Reviewer: that's already possible with current agents though no? patterns and conventions are unrelated to next

Author: Not really, the model is pretty behind, like sonnet's knowledge is from Jan 2025, it doesn't have enough knowledge of new APIs and even makes a lot mistakes as well. hallucination with next APIs is very common even unless you force the agent to search the docs first for specific case.

Some guides and knowledge is built into `next-devtools-mcp` to help change that

Reviewer: let's differentiate between next.js patterns and converntions and the user ones then

## Development workflow

1. **Start your Next.js server** with the MCP server enabled:

```bash filename="Terminal"
npm run dev
```

2. **Connect your Coding Agent** to the running Next.js instance

> 3. **Load your application in the browser** to trigger compilation and rendering. The MCP tools work by analyzing the compiled and rendered state of your application.

Reviewer: last part of the phrase about compiling etc can be cut off.

4. **Query your agent** for insights and diagnostics:

```

> What errors are currently in my application?

Reviewer: is that the list of _all_ possible commands? we really need the docs to list eveyrthing. otherwise it's too vague.

Reviewer 2: +1

---
title: Enabling Next.js MCP Server for Coding Agents
nav_title: Next.js MCP Server
description: Learn how to enable the experimental Next.js MCP server to allow agents access your application state
related:
  title: Learn more about MCP
  links:
    - app/guides/mcp
---

The Model Context Protocol (MCP) is an open standard that allows AI agents and coding assistants to interact with your applications through a standardized interface. The experimental Next.js MCP server enables agents to understand and work with your application's structure and state in real-time.

When combined with [`next-devtools-mcp`](https://www.npmjs.com/package/next-devtools-mcp), your agent gets access to comprehensive Next.js development tools including documentation, browser testing capabilities, Chrome DevTools functionality, and automated MCP tools for common development tasks.

## Enabling the MCP server

To enable the experimental MCP server in your Next.js application, add the following to your `next.config.js`:

```js filename="next.config.js"
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    mcpServer: true,
  },
}

module.exports = nextConfig
```

> ### Configuring Next DevTools MCP

Reviewer 3: also say that **if** agent starts the dev server, then this step is optional

When you enable the MCP server, configure [`next-devtools-mcp`](https://www.npmjs.com/package/next-devtools-mcp) in a `.mcp.json` file at the root of your project to enhance the agent's capabilities with additional development tools:

```json filename=".mcp.json"
{
  "mcpServers": {
    "next-devtools": {
      "command": "npx",
      "args": ["-y", "next-devtools-mcp@latest"]
    }
  }
}
```

See the [next-devtools-mcp documentation](https://www.npmjs.com/package/next-devtools-mcp) for more configuration options and features.

## What the MCP server provides

With `experimental.mcpServer` enabled and `next-devtools-mcp` configured, agents get access to:

- **Next.js Knowledge Base** - Comprehensive Next.js documentation and best practices
- **Debugging Insights** - Real-time error detection, build diagnostics, logs, and runtime information from your dev server
- **Migration Tools** - Automated helpers for migrating your application, including upgrading to Next.js 16 with codemods and enabling Cache Components with automated setup and error detection

## Using with agents

> When you have the MCP server enabled, MCP-compatible agents (like Claude Code) can automatically discover and connect to your Next.js application with the experimental `mcpServer` flag enabled. This provides agents with rich context about your project.

Reviewer 3: prefer not to mention "claude code" if we can

---
title: Enabling Next.js MCP Server for Coding Agents
nav_title: Next.js MCP Server
description: Learn how to enable the experimental Next.js MCP server to allow agents access to your application state
---

The [Model Context Protocol (MCP)](https://modelcontextprotocol.io) is an open standard that allows AI agents and coding assistants to interact with your applications through a standardized interface.

## Two complementary MCP servers

> We recommend using both MCP servers together for the best agent experience:

Reviewer 2: ```suggestion
Next.js provides two MCP servers, we recommend using both together for the best agent experience:
```

### 1. Next.js MCP Server (Built-in)

**Available in Next.js 16 and above**

The Next.js MCP server exposes your application's internals, allowing agents to:

- Access real-time application state and runtime information
- Query page metadata, routes, and rendering details
- Retrieve build errors, runtime errors, and development logs
- Inspect Server Actions and component hierarchies

This server runs within your Next.js dev server and provides direct access to your application's internal state.

### 2. Next DevTools MCP (External Package)

[`next-devtools-mcp`](https://www.npmjs.com/package/next-devtools-mcp) is a separate package that provides high-level development tools and guidance:

- **Next.js Knowledge Base** - Query comprehensive Next.js documentation and best practices

> - **Migration & Upgrade Tools** - Automated helpers for upgrading to Next.js 16 with codemods

Reviewer 2: ```suggestion
- **Migration and Upgrade Tools** - Automated helpers for upgrading to Next.js 16 with codemods
```

- **Cache Components Guide** - Setup and configuration assistance for Cache Components

> - **Browser Testing** - [Playwright MCP](https://github.com/microsoft/playwright-mcp) integration for verifying pages in the browser

Reviewer 2: ```suggestion
- **Next.js Knowledge Base**: Query comprehensive Next.js documentation and best practices
- **Migration & Upgrade Tools**: Automated helpers for upgrading to Next.js 16 with codemods
- **Cache Components Guide**: Setup and configuration assistance for Cache Components
- **Browser Testing**: [Playwright MCP](https://github.com/microsoft/playwright-mcp) integration for verifying pages in the browser
```

> Additionally, `next-devtools-mcp` can automatically discover and communicate directly with running Next.js dev servers to access their MCP tools (`get_errors`, `get_logs`, `get_page_metadata`, etc.). This integration allows agents to seamlessly use both high-level development guidance and low-level application internals to help you develop better.

Reviewer 2: Cut since it's repeated below. 

```suggestion
Additionally, `next-devtools-mcp` can automatically discover and communicate directly with running Next.js dev servers to access their MCP tools (`get_errors`, `get_logs`, `get_page_metadata`, etc.).
```

Together, these two MCP servers give agents both low-level application internals and high-level development guidance.

## Getting started

> The Next.js MCP server is enabled by default in Next.js v16 or above when running in development mode.

Reviewer 2: Nit, clarity 

```suggestion
The built-in Next.js MCP server is enabled by default in development in Next.js 16 or above.
```

### Configuring Next DevTools MCP

> To enhance the agent's capabilities with additional development tools, configure [`next-devtools-mcp`](https://www.npmjs.com/package/next-devtools-mcp) in a `.mcp.json` file at the root of your project:

Reviewer 2: ```suggestion
To use the Next Devtools MCP, add  [`next-devtools-mcp`](https://www.npmjs.com/package/next-devtools-mcp) to the `.mcp.json` file at the root of your project
```

```json filename=".mcp.json"
{
  "mcpServers": {
    "next-devtools": {
      "command": "npx",
      "args": ["-y", "next-devtools-mcp@latest"]
    }
  }
}
```

See the [next-devtools-mcp documentation](https://www.npmjs.com/package/next-devtools-mcp) for more configuration options and features.

## Using with agents

MCP-compatible Coding Agents can automatically discover and connect to your Next.js development server. This provides agents with rich context about your project.

### Benefits for agent-assisted development

Agents can:

- **Make context-aware suggestions** - Recommend the right place to add new features based on your existing structure
- **Query live application state** - Check current configuration, routes, and middleware during development
- **Understand your app router pages layout** - Know exactly which page and layout are rendered in the page

> - **Provide accurate implementations** - Generate code that follows your project's patterns and conventions

Reviewer 2: ```suggestion
- **Make context-aware suggestions**:  Recommend the right place to add new features based on your existing structure
- **Query live application state**: Check current configuration, routes, and middleware during development
- **Understand your app router pages layout**: Know exactly which page and layout are rendered in the page
- **Provide accurate implementations**: Generate code that follows your project's patterns and conventions
```

## Development workflow

1. **Start your Next.js development server**:

> 

Reviewer 2: ```suggestion
1. Start your Next.js development server:
```

```bash filename="Terminal"
npm run dev
```

> 2. **Connect your Coding Agent** to the running Next.js instance

Reviewer 2: ```suggestion
2. Connect your Coding Agent to the running Next.js instance
```

> 3. **Load your application in the browser** to load the page.

Reviewer 2: ```suggestion
3. Load your application in the browser to load the page.
```

> 4. **Query your agent** for insights and diagnostics:

Reviewer 2: ```suggestion
4. Query your agent for insights and diagnostics:
```

> ```

Reviewer 4: I think we can change to:

````suggestion
```bash filename="Terminal"
What errors are currently in my application?
```
````

What errors are currently in my application?
```

The agent will:

- Call the `get_errors` tool from the Next.js MCP server
- Get real-time error information, logs, and build diagnostics from your running dev server
- Provide actionable fixes based on the actual state of your compiled application

### Available Next.js MCP Tools

The experimental Next.js MCP server provides the following tools:

- **get_errors** - Retrieve current build errors, runtime errors, and type errors from your dev server
- **get_logs** - Access development server logs and console output
- **get_page_metadata** - Get metadata about specific pages including routes, components, and rendering information
- **get_project_metadata** - Retrieve project structure, configuration, and overall metadata

> - **get_server_action_by_id** - Look up Server Actions by their ID for debugging and inspection

Reviewer 2: ```suggestion
- **get_errors**: Retrieve current build errors, runtime errors, and type errors from your dev server
- **get_logs**: Access development server logs and console output
- **get_page_metadata**: Get metadata about specific pages including routes, components, and rendering information
- **get_project_metadata**: Retrieve project structure, configuration, and overall metadata
- **get_server_action_by_id**: Look up Server Actions by their ID for debugging and inspection
```

> 5. **Use migration and knowledge tools** from next-devtools-mcp:

Reviewer 2: Step 5 is out of the flow (broken by the **Available MCP tools** heading), I'd call this examples. 
```suggestion
Examples: 
```

```
Help me upgrade my Next.js app to version 16
```

or

```
When should I use "use client" in App Router?
```

The agent will use next-devtools-mcp to guide you through automated migrations with codemods and query the Next.js documentation for best practices.

## Example: Error detection and diagnosis

The agent can query your running Next.js application for real-time error information:

```
User: "What errors are currently in my application?"

Agent action:
- Calls the `get_errors` tool provided by the Next.js MCP server
- Retrieves current build errors, runtime errors, and type errors from the dev server
- Analyzes the errors and provides actionable fixes

Agent response:
- Lists all errors with detailed explanations
- Suggests code changes to fix each error
- Prioritizes errors by severity and impact
- Recommends best practices to prevent similar issues
```

> This demonstrates the power of having both MCP servers enabled: the Next.js MCP server provides real-time diagnostics directly from your application, while next-devtools-mcp can provide guidance on best practices and migration paths.

Reviewer 2: This section is a duplicate from above, you could remove it or combine all into one example.

- Retrieve build errors, runtime errors, and development logs
- Inspect Server Actions and component hierarchies

This server runs within your Next.js dev server and provides direct access to your application's internal state.

### 2. Next DevTools MCP (External Package)

[`next-devtools-mcp`](https://www.npmjs.com/package/next-devtools-mcp) is a separate package that provides high-level development tools and guidance:

- **Next.js Knowledge Base**: Query comprehensive Next.js documentation and best practices
- **Migration and Upgrade Tools**: Automated helpers for upgrading to Next.js 16 with codemods
- **Cache Components Guide**: Setup and configuration assistance for Cache Components
- **Browser Testing**: [Playwright MCP](https://github.com/microsoft/playwright-mcp) integration for verifying pages in the browser

Additionally, `next-devtools-mcp` can automatically discover and communicate directly with running Next.js dev servers to access their MCP tools (`get_errors`, `get_logs`, `get_page_metadata`, etc.).

Together, these two MCP servers give agents both low-level application internals and high-level development guidance.

## Getting started

> The built-in Next.js MCP server is enabled by default in development in Next.js 16 or above.

Reviewer 4: @huozhi the bot is pulling from older data right?

> The built-in Next.js MCP server is enabled by default in development in Next.js 16 or above.

Author: We can ignore this, it's enabled on canary by default

- Analyzes the errors and provides actionable fixes

<details>
<summary>Agent response (click to expand)</summary>

```
> fix errors on the page

⏺ I'll help you fix the errors on the page. Let me first check what errors exist by looking at the
  current state of the project.

⏺ next-devtools - nextjs_runtime (MCP)(action: "discover_servers")
⏺ next-devtools - nextjs_runtime (MCP)(action: "call_tool", toolName: "get_errors")

⏺ I found a hydration error on the /about page. The error shows that the server is rendering
  "server" but the client is rendering "client", causing a mismatch. Let me create a todo list and fix
   this issue.

⏺ Perfect! Now let me check for errors on the page.           01:14 PM  claude-sonnet-4-5-20250929

⏺ next-devtools - nextjs_runtime (MCP)(action: "call_tool", toolName: "get_errors")
  ⎿ {
      "success": true,
      "port": 3000,
      "toolName": "get_errors",
      "result": {
        "content": [
          {
            "type": "text",
            "text": "# Found errors in 1 browser session(s)\n\n## Session: /about\n\n**1 error(s)
     found**\n\n### Runtime Errors\n\n#### Error 1 (Type: recoverable)\n\n**Error**: Hydration failed
```

> </details>

Author: it's very long we have to truncate
